Initial revisions on the attacks section (#34)
xsleaks/wiki: content/docs/attacks/cache-probing.md

resources are cached it may be enough to leak whether a user accessed a specific page in the past.

A variation on cache probing abuses [Error Events]({{< ref "../attacks/error-events.md" >}}) to perform more accurate and impactful attacks.

## Attack Principle

An attacker wants to know whether a user visited a certain social network.

1. A user visits a social network and some of the subresources will be cached. 
2. The user visits an attacker controlled page which will fetch a resource that is usually fetched by that social network. 
3. Using a [Network Timing XS-Leak]({{< ref "../attacks/timing-attacks/network-timing.md" >}}), the attacker page can detect the difference from a response coming from the cache (step 1 happened) or coming from the network (step 1 did not happen). The delay will be significantly lower in a request served from the cache.

## Cache Probing with Error Events

> Cache Probing with [Error Events]({{< ref "../attacks/error-events.md" >}}) [^2] allows more accurate. Instead of relying on timing measurements, this leverages [Error Events]({{< ref "../attacks/error-events.md" >}}) and some server-side behavior to detect whether a resource was cached. The attack works as follows:

Reviewer: ```suggestion
Cache Probing with [Error Events]({{< ref "../attacks/error-events.md" >}}) [^2] allows more accurate attacks. Instead of relying on timing measurements, this leverages [Error Events]({{< ref "../attacks/error-events.md" >}}) and some server-side behavior to detect whether a resource was cached. The attack works as follows:
```

1. [Invalidate the resource]({{< ref "#invalidate-the-cache" >}}) from the browser cache. This step is required to make sure the attack will not consider a resource previously cached in another visit.
2. Perform a request that will cause different items to be cached depending on the user's state. For example, load a page that will include a specific image only if the user is logged in. This request can be triggered by navigating to the target website with `<link rel=prerender..`, embedding the website in an `iframe`, or opening a new window with `window.open`.
3. Trigger a request that will cause the server to reject the request. For example, include an [overly long referer header](https://lists.archive.carbon60.com/apache/users/316239) that will make the server reject the request. If the resource was cached in step 2, this request will succeed instead of triggering an error event. 

### Invalidate the cache

To invalidate a resource from the cache the attacker must force the server to return an error when fetching that subresource. There are a couple of ways to achieve this:

> - A request with an [overly long referer header](https://lists.archive.carbon60.com/apache/users/316239) and `'cache':'reload'`. This might not work as browsers [capped](https://github.com/whatwg/fetch/issues/903) the length of the referrer to prevent this.

Reviewer: Is `overly long` commonly used construct to say `overlong`? I think I only saw `overlong headers` in that context. What's wrong with `overlong`?

Author: You're right! Personally I don't think I've ever heard/used the word overlong but it is totally a valid and well used word. Reverted this.

- A `POST` request with a `fetch` `no-cors`. Sometimes even in cases where an error is not returned the browser invalidates the cache.
- Request Headers such as Content-Type, Accept, Accept-Language, etc that may cause the server to fail (more application dependent).
- Other request properties.

> Often some of these methods might be considered a bug in the browser (ie [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=959789#c9)). 

Reviewer: ```suggestion
Often some of these methods might be considered a bug in the browser (e.g. [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=959789#c9)). 
```

|:---------------------------------:|:-------------------------------------:|:---------------------------------------:|
|        ✔️ [(if Strict)]({{< ref "../defenses/opt-in/same-site-cookies.md#lax-vs-strict" >}})             |                  ✔️                   |   ✔️ & ❌ [*]({{< ref "../defenses/design-protections/subresource-protections.md#deployment" >}})   | 

- If a resource can be fetched with user authentication, [Same-Site Cookies]({{< ref "../defenses/opt-in/same-site-cookies.md" >}}) (Strict) can protect that resource from being abused by an attacker origin.
- [Vary: Sec-Fetch-Site]({{< ref "../defenses/opt-in/fetch-metadata.md#fetch-metadata--cache-probing-attacks" >}}) allows applications to force cache segregation by a group of origins.
- [Subresource Protections]({{< ref "../defenses/design-protections/subresource-protections.md" >}}) allow application to set random tokens in URLs to make them unpredictable from attackers. Useful for both authenticated and unauthenticated subresources.

### Default Defense Mechanisms

> [Paritioned Caches]({{< ref "../defenses/browser-default/partitioned-cache.md" >}}) are a feature implemented in browsers to create a unique cache for each origin. This protection prevents an attacker's origin from interfering with cached resources of other origins. Applications do not have to opt-in to enforce this.

Reviewer: ```suggestion
[Paritioned HTTP Cache]({{< ref "../defenses/browser-default/partitioned-cache.md" >}}) is a feature implemented in browsers to create a unique cache for each origin. This protection prevents an attacker's origin from interfering with cached resources of other origins. Applications do not have to opt-in to enforce this.
```

Author: Changed in a commit locally while fixing another typo
